test/screenshot/lib/snapshot-store.js

```diff
@@ -117,6 +131,97 @@ class SnapshotStore {
     const rawDiffBase = this.cliArgs_.diffBase;
     throw new Error(`Unable to parse '--mdc-diff-base=${rawDiffBase}': Expected a URL, local file path, or git ref`);
   }
+
+  /**
+   * @param {!Array<!UploadableTestCase>} testCases
+   * @param {!Array<!ImageDiffJson>} diffs
+   * @return {!Promise<!SnapshotSuiteJson>}
+   * @private
+   */
+  async getJsonData_({testCases, diffs}) {
+    return this.cliArgs_.hasAnyFilters()
+      ? await this.updateFilteredScreenshots_({testCases, diffs})
+      : await this.updateAllScreenshots_({testCases, diffs});
+  }
+
+  /**
+   * @param {!Array<!UploadableTestCase>} testCases
+   * @param {!Array<!ImageDiffJson>} diffs
+   * @return {!Promise<!SnapshotSuiteJson>}
+   * @private
+   */
+  async updateFilteredScreenshots_({testCases, diffs}) {
+    const oldJsonData = await this.fromDiffBase();
+    const newJsonData = await this.fromTestCases(testCases);
+    const jsonData = this.deepCloneJson_(oldJsonData);
+
+    diffs.forEach((diff) => {
+      const htmlFilePath = diff.htmlFilePath;
+      const browserKey = diff.browserKey;
+      if (jsonData[htmlFilePath]) {
+        jsonData[htmlFilePath].publicUrl = newJsonData[htmlFilePath].publicUrl;
+        jsonData[htmlFilePath].screenshots[browserKey] = newJsonData[htmlFilePath].screenshots[browserKey];
+      } else {
+        jsonData[htmlFilePath] = this.deepCloneJson_(newJsonData[htmlFilePath]);
+      }
+    });
+
+    return jsonData;
+  }
+
+  /**
+   * @param {!Array<!UploadableTestCase>} testCases
+   * @param {!Array<!ImageDiffJson>} diffs
+   * @return {!Promise<!SnapshotSuiteJson>}
+   * @private
+   */
+  async updateAllScreenshots_({testCases, diffs}) {
+    const oldJsonData = await this.fromDiffBase();
+    const newJsonData = await this.fromTestCases(testCases);
+
+    const existsInOldJsonData = ([htmlFilePath]) => htmlFilePath in oldJsonData;
+
+    /** @type {!Array<[string, !SnapshotPageJson]>} */
+    const newMatchingPageEntries = Object.entries(newJsonData).filter(existsInOldJsonData);
+
+    for (const [htmlFilePath, newPage] of newMatchingPageEntries) {
+      let pageHasChanges = false;
+
+      for (const browserKey of Object.keys(newPage.screenshots)) {
+        const oldUrl = oldJsonData[htmlFilePath].screenshots[browserKey];
+        const matchingDiff = diffs.find((diff) => diff.htmlFilePath === htmlFilePath && diff.browserKey === browserKey);
+
+        if (oldUrl && !matchingDiff) {
+          newPage.screenshots[browserKey] = oldUrl;
+        }
+
+        if (matchingDiff) {
+          pageHasChanges = true;
+        }
+      }
+
+      if (!pageHasChanges) {
+        newPage.publicUrl = oldJsonData[htmlFilePath].publicUrl;
+      }
+    }
+
+    return newJsonData;
```

Comment:
Maybe I'm missing something but where does `newJsonData` get modified in the above loops?

Reply:
It might not be 100% obvious, but the `for` loops iterate over `newJsonData` and change some of its properties:

```js
newPage.screenshots[browserKey] = oldUrl;
...
newPage.publicUrl = oldJsonData[htmlFilePath].publicUrl;
```

I tried to name all the variables `newPage` and `newMatchingPageEntries` to hint that they're from `newJsonData`, but I maybe it's still not totally clear 😕

Any advice?

Reply:
How is it changing the properties? The variable you're updating is `newPage`. Is `newPage` a copy of `newJsonData`?

Reply:
`newPage` is a pointer to a sub-object within `newJsonData`.

I basically loop over the pages and screenshots in `newJsonData`, and if I find a screenshot that hasn't changed, I copy the URL from `oldJsonData` and set it on `newPage` (which is just a reference to an object in `newJsonData`).

E.g.:

Here's `newJsonData`:
```json
{
  "page.html": { // <-- `newPage`
    "publicUrl": "NEW/page.html",
    "screenshots": {
      "desktop_windows_ie@11": "NEW/page.html.ie11.png"
    }
  }
}
```

Here's `oldJsonData`:
```json
{
  "page.html": {
    "publicUrl": "OLD/page.html",
    "screenshots": {
      "desktop_windows_ie@11": "OLD/page.html.ie11.png"
    }
  }
}
```

Does that make sense at all?

(Sorry for the confusion... I knew this method might be kinda hard to understand, but I wanted to get something working ASAP.)

Reply:
Ah, I see. Yeah, that's not super clear. Since you've explained it, I get it now, so I'll stamp. That being said, we should make an issue to clear this up soon. This is one of the more important parts of screenshot testing and should be pretty clear.

Reply:
K, I added a TODO to refactor this method. Thanks!